README.md
# CloudWatch Postman [![Codeship Status for KissKissBankBank/cloudwatch-postman](https://app.codeship.com/projects/095c5ef0-5d45-0137-3423-3220a1e64d22/status?branch=master)](https://app.codeship.com/projects/343038)

CloudWatch Postman is a Node server that sends data to [Amazon
CloudWatch](https://aws.amazon.com/cloudwatch/). It enables you to serve an API
with endpoints that add or update your metrics/your logs on CloudWatch with
your AWS credentials.

- For the moment, configuration of this API is set with environment variables.
- Its first purpose is to serve some endpoints so an cliend-side application can
  call it to send data to CloudWatch.

![Postman on a bicycle](https://user-images.githubusercontent.com/548778/57973195-5be7e680-79a5-11e9-9422-a8e59faa8384.gif)

- [Prerequisites](#prerequisites)
- [Quick start](#quick-start)
- [How to request the API](#how-to-request-the-api)
  - [What is a unique access token?](#what-is-a-unique-access-token)
  - [When is a unique access token used?](#when-is-a-unique-access-token-used)
  - [How to get a unique access token?](#how-to-get-a-unisque-access-token)
  - [How to generate your client token](#how-to-generate-your-client-token)
- [API](#api)
- [Configuration with Dotenv](#configuration-with-dotenv)
- [Contributing](#contributing)
- [Resources](#resources)

## Prerequisites

- Node.js,
- Redis,
- Amazon CloudWatch,
- an AWS IAM account that can call CloudWatch with read and write access.

## Quick start

Install the dependencies:

```sh
npm install
```

Choose an `CLIENT_SECRET_KEY` and an `ACCESS_TOKEN_SECRET_KEY`. These secret values
will be used by CloudWatch Postman to generate tokens to access the API.

[Create a `.env` file](#configuration-with-dotenv) with these secrets and your
AWS credentials.

Start the app:

```sh
npm run serve
```

Test the API on [http://localhost:8080/test](http://localhost:8080/test).

## How to request the API

You can request the API using a unique access token:

```
+--------------------+                               +--------------------+
|                    | 2. Ask for an access token.   |                    |
|                    | +------- POST /token -------> |                    |
|                    |                               |                    |
|       Client       | <---------------------------+ |                    |
| (your application) | 3. The API returns a valid    |      The API       |
|                    |    access token.              |                    |
|                    |                               | cloudwatch-postman |
|  1. Create your    |                               |                    |
|     client token.  |                               |                    |
|                    |                               |                    |
|                    |                               |                    |
|                    |                               |                    |
|                    |                               |                    |
|                    | 4. Make API calls with the    |                    |
|                    |    valid access token.        |                    |
|                    |                               |                    |
|                    | +---- eg. POST /metric -----> |                    |
|                    |                               |                    |
|                    | (by default, the access token |                    |
|                    |  is valid for one hour)       |                    |
+--------------------+                               +--------------------+
```

### What is a unique access token?

As CloudWatch Postman is firstly meant to be called by a client-side
application, unique access tokens can secure a little bit more the API endpoints.

You need to exchange your [client token](#how-to-generate-your-client-token) to
obtain a unique access token. The latter have a default expiration of one
hour.

### When is a unique access token used?

Every endpoint, except the `POST /token` one, needs an `accessToken` to be
requested. We advise you to fetch it on your client-side application as soon as
possible if you know that you will need to query the API.

### How to get a unique access token?

You can fetch an `accessToken` on the `POST /token` endpoint with your
[client token](#how-to-generate-your-client-token).

### How to generate your client token

This section explains how to generate an client token to request an access
token for the API.

You will needs these values:
- the current timestamp,
- a random salt value,
- your `CLIENT_SECRET_KEY`,

Concatenate these 3 values and hash them with a `sha256` algorithm digested in
`base64`. Here is an
example in JavaScript:
```js
import crypto from 'crypto'

const data = `${timestamp}${salt}${appSecretKey}`
const hash = crypto.createHash('sha256').update(data).digest('base64')
```

Then, generate your token:
- concatenate the date, the salt value and the hash with a `::` delimiter,
- encode this string in `base64`.

Here is an example in JavaScript:
```js
Buffer.from([timestamp, salt, hash].join('::')).toString('base64')
```

### Tokens usage

Name | Usage | How to get it | Expiration
--- | --- | ---  | ---
  Access token | An access token is used for each call to the API endpoints, except `POST /token` and `GET /test`. It has to be included in the JSON body of your calls along with your other parameters `{ accessToken: "aValidAccessToken" }`. | You can fetch a valid access token on `POST /token` with your client token. | By default, an access token has a one hour validity from the moment it is sent to the client.
Client token | The client token is used to fetch a valid access token on `POST /token`. It has to be included in the JSON body of your call `{ clientToken: 'yourClientToken'}`. | You have to [generate your client token](#how-to-generate-your-client-token) on your side. | By default, a client token has a one day validity from the moment it is generated on your side.

## API

You can check the existing endpoints of this API in [the
documentation](https://github.com/KissKissBankBank/cloudwatch-postman/tree/master/docs/api.md).


## Configuration with Dotenv

You can set some variables with a `.env` file and start the app with:

```sh
npm run serve
```

### Example

```sh
# .env
AWS_ACCESS_KEY_ID=***
AWS_SECRET_ACCESS_KEY=***
AWS_REGION=***
CLIENT_SECRET_KEY=***
ACCESS_TOKEN_SECRET_KEY=***
```

### Variables

The following variables can be setup in the `.env` file:

Variable | Requirement | Description | Default value
--- | --- | ---  | ---
`AWS_ACCESS_KEY_ID` | *Required* | The AWS IAM user access key id. |
`AWS_SECRET_ACCESS_KEY` | *Required* | The AWS IAM user secret access key. |
`AWS_REGION` | *Required* | The CloudWatch region |
`CLIENT_SECRET_KEY` | *Required* | Your client secret key. You will share it on your consumer app to generate your client token. |
`ACCESS_TOKEN_SECRET_KEY` | *Required* | Your access token secret key. It is used to generate all the access tokens. |
`PORT` | *Optional* | The port on which the server is lauched | `8080`
`CORS_ALLOWED_ORIGIN` | *Optional* | A list of domain origins to which you grant the access to your API. Separate each origin with a comma: `alice-in-wonderland.io, the-mad-hatter.com, tweedledee-tweedled.um` | `*`

## Contributing

Please refer to the [contributing
documentation](https://github.com/KissKissBankBank/cloudwatch-postman/tree/master/docs/contributing.md).

## Resources

- [AWS IAM user policies for CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/iam-identity-based-access-control-cwl.html)
- [AWS IAM user credentials](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/getting-your-credentials.html)
- [Amazon CloudWatch API for putMetricData](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudWatch.html#putMetricData-property)
- [Amazon CloudWatch Logs API for
  putLogEvents](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html)


CHANGELOG.md
# Changelog

This project adheres to [Semantic Versioning](http://semver.org/).

## [unreleased]

##
[4.0.2](https://github.com/KissKissBankBank/cloudwatch-postman/compare/v4.0.1...v4.0.2) - 2019-07-02

Fix:
- [#30](https://github.com/KissKissBankBank/cloudwatch-postman/pull/30) - Clean
  and reuse Redis connections when possible

## [4.0.1](https://github.com/KissKissBankBank/cloudwatch-postman/compare/v4.0.0...v4.0.1) - 2019-06-12

Fix:
- [#27](https://github.com/KissKissBankBank/cloudwatch-postman/pull/27) - Use
  `REDIS_URL` environment variable to instanciate Bull queues

## [4.0.0](https://github.com/KissKissBankBank/cloudwatch-postman/compare/v3.0.0...v4.0.0) - 2019-06-12

Breaking changes:
- [#21](https://github.com/KissKissBankBank/cloudwatch-postman/pull/21) - Change
  the configuration variable `APP_SECRET_KEY` into `CLIENT_SECRET_KEY`
- [#24](https://github.com/KissKissBankBank/cloudwatch-postman/pull/24) - Update
  `POST /logEvents` endpoint:
  - Delegate sequenced calls to CloudWatch Logs to a dedicated worker,
  - Remove response body,
  - Change request parameters. You should pass the following parameters now:
  ```json
  {
    "accessToken": "YOUR_UNIQUE_ACCESS_TOKEN",
    "logGroupName": "YOUR_LOG_GROUP_NAME",
    "logStreamName": "YOUR_LOG_STREAM_NAME",
     "logEvents": [{
      "message": "LOG MESSAGE",
      "timestamp": "TIMESTAMP",
    }]
  }
  ```

Features:
- [#20](https://github.com/KissKissBankBank/cloudwatch-postman/pull/20) - Add
  CORS handling on API endpoints
- [#26](https://github.com/KissKissBankBank/cloudwatch-postman/pull/26) - Add a
  `Procfile` to start the application


## [3.1.0](https://github.com/KissKissBankBank/cloudwatch-postman/compare/v3.0.0...v3.1.0) - 2019-05-23

- [#18](https://github.com/KissKissBankBank/cloudwatch-postman/pull/18) - Add an endpoint `POST /logEvents` to create log events in CloudWatch

Feature:
- [#18](https://github.com/KissKissBankBank/cloudwatch-postman/pull/18) - Add an endpoint `POST /logEvents` to create log events in CloudWatch

## [3.0.0](https://github.com/KissKissBankBank/cloudwatch-postman/compare/v2.0.0...v3.0.0) - 2019-05-20

Breaking changes:
- [#11](https://github.com/KissKissBankBank/cloudwatch-postman/pull/11)/[#12](https://github.com/KissKissBankBank/cloudwatch-postman/pull/12) - Remove the application token authorization on `GET /test` endpoint.
- [#11](https://github.com/KissKissBankBank/cloudwatch-postman/pull/11)/[#12](https://github.com/KissKissBankBank/cloudwatch-postman/pull/12) - Add rate limit by IP on server endpoints. All endpoints follows these options for `restify-throttle`:
```js
{
  burst: 3,  // Max 3 concurrent requests
  rate: 1,  // 1 request / second
  ip: true,   // throttle per IP
}
```

The `GET /test` endpoint has a specific throttle:
```js
{
  burst: 1,
  rate: 0.015, // 1 request / minute
  ip: true,
}
```

Feature:
- [#8](https://github.com/KissKissBankBank/cloudwatch-postman/pull/8) - Update token creation strategy

## [2.0.0](https://github.com/KissKissBankBank/cloudwatch-postman/compare/v1.0.0...v2.0.0) - 2019-05-20

Breaking change:
- [#7](https://github.com/KissKissBankBank/cloudwatch-postman/pull/7) - Add an authorization with an access token to request POST /metric

Feature:
- [#7](https://github.com/KissKissBankBank/cloudwatch-postman/pull/7) - Add POST /token endpoint to enable client to request an access token

## 1.0.0 - 2019-05-19

- First release!


test/index.js
import chai, { expect } from 'chai'
import chaiHttp from 'chai-http'
import sinon from 'sinon'
import sinonChai from 'sinon-chai'
import server from './../lib/index'
import AWS from 'aws-sdk-mock'
import {
  generateSalt,
  generateHash,
  generateToken,
  createAccessToken,
} from '../lib/token'

const Queue = require('bull')

global.chai = chai
global.expect = expect
chai.use(chaiHttp)
chai.use(sinonChai)

export const createTestClientToken = () => {
  const date = new Date().getTime()
  const salt = generateSalt(8)
  const secret = process.env.CLIENT_SECRET_KEY
  const hash = generateHash(`${date}${salt}${secret}`)

  return generateToken([date, salt, hash])
}

describe('=========== index.js ==========', () => {
  beforeEach(() => {
    AWS.mock('CloudWatch', 'putMetricData', function (params, callback){
      callback(null, { ResponseMetadata: { RequestId: 'foobar' }})
    });
  })

  afterEach(() => {
    AWS.restore('CloudWatch')
  })

  describe('GET /test', () => {
    it('returns a 200 with a data object', (done) => {
      chai.request(server)
        .get(`/test`)
        .end((err, res) => {
          const { status, body } = res

          expect(status).to.eq(200)
          expect(body.ResponseMetadata.RequestId).to.eq('foobar')

          done()
        })
    })
  })

  describe('POST /token', () => {
    describe('with a valid app token', () => {
      it('returns a 201 with the newly created access token', (done) => {
        const clientToken = createTestClientToken()

        chai.request(server)
          .post('/token')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify({ clientToken }))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(201)
            expect(body).to.have.key('accessToken')

            done()
          })
      })
    })

    describe('with an invalid app token', () => {
      it('returns a 401 with an error object', (done) => {
        const clientToken = 'aliceInWonderland'

        chai.request(server)
          .post('/token')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify({ clientToken }))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(401)
            expect(body.error.code).to.eq(101)
            expect(body.error.message)
              .to.eq('Required parameter "clientToken" is invalid.')

            done()
          })
      })
    })

    describe('without an app token', () => {
      it('returns a 403 with an error object', (done) => {
        chai.request(server)
          .post('/token')
          .set('Content-Type', 'application/json')
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(403)
            expect(body.error.code).to.eq(100)
            expect(body.error.message)
              .to.eq('Required parameter is missing: "clientToken".')

            done()
          })
      })
    })
  })

  describe('POST /metric', () => {
    describe('with a valid access token', () => {
      it('returns a response', (done) => {
        const accessToken = createAccessToken()
        const params = {
          accessToken,
          params: {
            MetricData: [
              {
                MetricName: 'HELLO_WORLD',
                Value: 100,
              }
            ],
            Namespace: 'cloudwatch-postman',
          }
        }

        chai.request(server)
          .post('/metric')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(params))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(201)
            expect(body.ResponseMetadata.RequestId).to.eq('foobar')

            done()
          })
      })
    })

    describe('with an invalid access token', () => {
      it('returns a 401 with an error object', (done) => {
        const accessToken = 'aliceInWonderland'

        const params = {
          accessToken,
          params: {
            MetricData: [
              {
                MetricName: 'HELLO_WORLD',
                Value: 100,
              }
            ],
            Namespace: 'cloudwatch-postman',
          }
        }

        chai.request(server)
          .post('/metric')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(params))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(401)
            expect(body.error.code).to.eq(111)
            expect(body.error.message)
              .to.eq('Required parameter "accessToken" is invalid.')

            done()
          })
      })
    })

    describe('without an access token', () => {
      it('returns a 403 with an error object', (done) => {
        const params = {
          params: {
            MetricData: [
              {
                MetricName: 'HELLO_WORLD',
                Value: 100,
              }
            ],
            Namespace: 'cloudwatch-postman',
          }
        }

        chai.request(server)
          .post('/metric')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(params))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(403)
            expect(body.error.code).to.eq(110)
            expect(body.error.message)
              .to.eq('Required parameter is missing: "accessToken".')

            done()
          })
      })
    })
  })

  describe('POST /logEvents', () => {
    let sandbox

    before(() => sandbox = sinon.createSandbox())
    afterEach(() => sandbox.restore())

    describe('with a valid access token', () => {
      it('returns a response', (done) => {
        const stubbedAdd = sandbox.stub(Queue.prototype, 'add')
        const accessToken = createAccessToken()
        const params = {
          accessToken,
          logGroupName: 'RUM',
          logStreamName: 'cloudwatch-postman-test',
          logEvents: [
            {
              message: 'This is a test.',
              timestamp: '1558602946107',
            },
          ],
        }

        chai.request(server)
          .post('/logEvents')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(params))
          .end((err, res) => {
            const { status, body } = res

            expect(stubbedAdd).to.have.been.calledOnce
            expect(status).to.eq(201)

            done()
          })
      })
    })

    describe('with an invalid access token', () => {
      it('returns a 401 with an error object', (done) => {
        const accessToken = 'aliceInWonderland'

        const params = {
          accessToken,
          params: {
            logGroupName: 'RUM',
            logStreamName: 'cloudwatch-postman-test',
            logEvents: [
              {
                message: 'This is a test.',
                timestamp: '1558602946107',
              },
            ],
            sequenceToken: 'the-cheshire-cat',
          },
        }

        chai.request(server)
          .post('/logEvents')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(params))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(401)
            expect(body.error.code).to.eq(111)
            expect(body.error.message)
              .to.eq('Required parameter "accessToken" is invalid.')

            done()
          })
      })
    })

    describe('without an access token', () => {
      it('returns a 403 with an error object', (done) => {
        const params = {
          params: {
            logGroupName: 'RUM',
            logStreamName: 'cloudwatch-postman-test',
            logEvents: [
              {
                message: 'This is a test.',
                timestamp: '1558602946107',
              },
            ],
            sequenceToken: 'the-cheshire-cat',
          },
        }

        chai.request(server)
          .post('/logEvents')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(params))
          .end((err, res) => {
            const { status, body } = res

            expect(status).to.eq(403)
            expect(body.error.code).to.eq(110)
            expect(body.error.message)
              .to.eq('Required parameter is missing: "accessToken".')

            done()
          })
      })
    })
  })
})


test/token.js
import chai, { expect } from 'chai'
import sinon from 'sinon'
import sinonChai from 'sinon-chai'
import {
  generateSalt,
  generateHash,
  generateToken,
  isTokenExpired,
  isClientTokenValid,
  isAccessTokenValid,
  createAccessToken,
} from '../lib/token'
import { createTestClientToken } from './index'

global.chai = chai
global.expect = expect
chai.use(sinonChai)

const crypto = require('crypto')
describe('========== token.js ==========', () => {
  let sandbox

  before(() => sandbox = sinon.createSandbox())
  afterEach(() => sandbox.restore())

  describe('generateSalt', () => {
    it('returns a string', () => {
      const toStringSpy = sandbox.spy()
      const stubbedRandomBytes = sandbox
        .stub(crypto, 'randomBytes')
        .returns({ toString: toStringSpy })

      generateSalt(10)
      expect(stubbedRandomBytes).to.have.been.calledWith(10)
      expect(toStringSpy).to.have.been.calledWith('hex')
    })
  })

  describe('generateHash', () => {
    it('returns a string', () => {
      const digestSpy = sandbox.spy()
      const stubbedUpdate = sandbox
        .stub()
        .returns({ digest: digestSpy })
      const stubbedCreateHash = sandbox
        .stub(crypto, 'createHash')
        .returns({ update: stubbedUpdate })

      generateHash('Alice in Wonderland')
      expect(stubbedCreateHash).to.have.been.calledWith('sha256')
      expect(stubbedUpdate).to.have.been.calledWith('Alice in Wonderland')
      expect(digestSpy).to.have.been.calledWith('base64')
    })
  })

  describe('generateToken', () => {
    it('returns a string', () => {
      const toStringSpy = sandbox.spy()
      const stubbedFrom = sandbox
        .stub(Buffer, 'from')
        .returns({ toString: toStringSpy })

      generateToken(['alice', 'in', 'wonderland'], '-')
      expect(stubbedFrom).to.have.been.calledWith('alice::in::wonderland')
      expect(toStringSpy).to.have.been.calledWith('base64')
    })
  })

  describe('isTokenExpired', () => {
    describe('without expiration parameter', () => {
      describe('when token is expired', () => {
        it('returns true', () => {
          const today = new Date()
          const expiredTokenDate = new Date()
          expiredTokenDate.setDate(today.getDate() - 2)
          const expiredTokenTime = expiredTokenDate.getTime()

          expect(isTokenExpired(expiredTokenTime)).to.eq(true)
        })
      })

      describe('when token is valid', () => {
        it('returns true', () => {
          const validTokenDate = new Date()
          const validTokenTime = validTokenDate.getTime()

          expect(isTokenExpired(validTokenTime)).to.eq(false)
        })
      })
    })

    describe('with expiration parameter', () => {
      describe('when token is expired', () => {
        it('returns true', () => {
          const today = new Date()
          const expiredTokenDate = new Date()
          expiredTokenDate.setHours(today.getHours() - 2)
          const expiredTokenTime = expiredTokenDate.getTime()

          expect(isTokenExpired(expiredTokenTime, { hour: 1 })).to.eq(true)
        })
      })

      describe('when token is valid', () => {
        it('returns true', () => {
          const validTokenDate = new Date()
          const validTokenTime = validTokenDate.getTime()

          expect(isTokenExpired(validTokenTime, { hour: 1 })).to.eq(false)
        })
      })
    })
  })

  describe('isAccessTokenValid', () => {
    describe('when access token is expired', () => {
      it('returns false', () => {
        const createTestAccessToken = () => {
          const today = new Date()
          const expiredTokenDate = new Date()
          expiredTokenDate.setDate(today.getDate() - 2)
          const expiredTokenTime = expiredTokenDate.getTime()
          const salt = generateSalt(12)
          const secret = process.env.ACCESS_TOKEN_SECRET_KEY
          const hash = generateHash(`${expiredTokenTime}${salt}${secret}`)

          return generateToken([expiredTokenTime, salt, hash])
        }
        const token = createTestAccessToken()

        expect(isAccessTokenValid(token)).to.eq(false)
      })
    })

    describe('when access token is valid', () => {
      describe('with a correct secret', () => {
        it('returns true', () => {
          const token = createAccessToken()

          expect(isAccessTokenValid(token)).to.eq(true)
        })
      })

      describe('with a wrong secret', () => {
        it('returns false', () => {
          const createTestAccessToken = () => {
            const today = new Date()
            const tokenTime = today.getTime()
            const salt = generateSalt(12)
            const secret = 'wrong-secret'
            const hash = generateHash(`${tokenTime}${salt}${secret}`)

            return generateToken([tokenTime, salt, hash])
          }
          const token = createTestAccessToken()

          expect(isAccessTokenValid(token)).to.eq(false)
        })
      })

      describe('with tampered data', () => {
        it('returns false', () => {
          const token = 'alice-in-wonderland'

          expect(isAccessTokenValid(token)).to.eq(false)
        })

      })
    })
  })

  describe('isClientTokenValid', () => {
    describe('when access token is expired', () => {
      it('returns false', () => {
        const createExpiredAppAccessToken = () => {
          const today = new Date()
          const expiredTokenDate = new Date()
          expiredTokenDate.setDate(today.getDate() - 2)
          const expiredTokenTime = expiredTokenDate.getTime()
          const salt = generateSalt(8)
          const secret = process.env.CLIENT_SECRET_KEY
          const hash = generateHash(`${expiredTokenTime}${salt}${secret}`)

          return generateToken([expiredTokenTime, salt, secret])
        }
        const token = createExpiredAppAccessToken()

        expect(isClientTokenValid(token)).to.eq(false)
      })
    })

    describe('when access token is valid', () => {
      describe('with a correct secret', () => {
        it('returns true', () => {
          const token = createTestClientToken()

          expect(isClientTokenValid(token)).to.eq(true)
        })
      })

      describe('with a wrong secret', () => {
        it('returns false', () => {
        const createWrongAppAccessToken = () => {
          const today = new Date()
          const tokenTime = today.getTime()
          const salt = generateSalt(8)
          const secret = 'wrong-secret'
          const hash = generateHash(`${tokenTime}${salt}${secret}`)

          return generateToken([tokenTime, salt, secret])
        }
        const token = createWrongAppAccessToken()

          expect(isClientTokenValid(token)).to.eq(false)
        })
      })

      describe('with tampered data', () => {
        it('returns false', () => {
          const token = 'alice-in-wonderland'

          expect(isClientTokenValid(token)).to.eq(false)
        })
      })
    })
  })

  describe('createAccessToken', () => {

  })
})


test/worker.js
import chai, { expect } from 'chai'
import AWS from 'aws-sdk-mock'
import sinon from 'sinon'
import sinonChai from 'sinon-chai'
import {
  tokenKey,
  getValidSequenceToken,
  saveValidSequenceToken,
} from '../lib/worker'

global.chai = chai
global.expect = expect
chai.use(sinonChai)

const redis = require('ioredis')

describe('=========== worker.js ==========', () => {
  let sandbox

  before(() => sandbox = sinon.createSandbox())
  afterEach(() => sandbox.restore())
})


lib/token.js
import crypto from 'crypto'

export const DELIMITER = '::'
export const generateSalt = length => crypto.randomBytes(length).toString('hex')
export const generateHash = data => (
  crypto.createHash('sha256').update(data).digest('base64')
)
export const generateToken = (tokenItems) => {
  return Buffer
    .from(tokenItems.join(DELIMITER))
    .toString('base64')
}

export const isTokenExpired = (tokenTimestamp, expiration = { day: 1 }) => {
  const today = new Date()
  const tokenDate = new Date(tokenTimestamp)
  const tokenExpirationDate = new Date(tokenTimestamp)
  const { day, hour } = expiration

  if (day) {
    tokenExpirationDate.setDate(tokenDate.getDate() + day)
  }

  if (hour) {
    tokenExpirationDate.setHours(tokenDate.getHours() + hour)
  }

  return tokenExpirationDate.getTime() < today.getTime()
}

export const isAccessTokenValid = (token) => {
  const data = Buffer.from(token, 'base64')
  const decodedData = data.toString('ascii')
  const accessTokenSecret = process.env.ACCESS_TOKEN_SECRET_KEY

  const [ date, salt, hash ] = decodedData.split(DELIMITER)

  if (isTokenExpired(+date, { hour: 1 })) return false

  const apiHash = generateHash(`${date}${salt}${accessTokenSecret}`)

  return apiHash == hash
}

export const isClientTokenValid = (token) => {
  const data = Buffer.from(token, 'base64')
  const decodedData = data.toString('ascii')
  const apiSecret = process.env.CLIENT_SECRET_KEY

  const [ date, salt, hash ] = decodedData.split(DELIMITER)

  if (isTokenExpired(+date)) return false

  const apiHash = generateHash(`${date}${salt}${apiSecret}`)

  return apiHash == hash
}

export const createAccessToken = () => {
  const date = new Date().getTime()
  const salt = generateSalt(12)
  const secret = process.env.ACCESS_TOKEN_SECRET_KEY
  const hash = generateHash(`${date}${salt}${secret}`)

  return generateToken([date, salt, hash])
}


lib/worker.js
import Queue from 'bull'
import { processor, logEventsQueueOptions } from './log-events-processor'

const logEventsQueue = new Queue('logEvents', logEventsQueueOptions)

logEventsQueue.process(processor)

logEventsQueue.on('completed', (job, result) => {
  console.log(
    '\n[cloudwatch-postman-worker] ' +
    'POST /putLogEvents to CloudWatch Logs completed. Response:',
    result,
  )
})

logEventsQueue.on('failed', (job, error) => {
  console.log(
    '\n[cloudwatch-postman-worker] ' +
    'POST /putLogEvents to CloudWatch Logs failed.',
    `\n${error}`,
  )
})

logEventsQueue.on('active', (job, jobPromise) => {
  console.log('\n[cloudwatch-postman-worker] A new job has started:')
  console.log(`ID: ${job.id}`)
  console.log('Data:', job.data)
})

console.log('\n[cloudwatch-postman-worker] logEventsQueue process is started!')


lib/index.js
import restify from 'restify'
import AWS from 'aws-sdk'
import request from 'request'
import {
  isClientTokenValid,
  isAccessTokenValid,
  createAccessToken,
} from './token'
import corsMiddleware from 'restify-cors-middleware'
import Queue from 'bull'
import { logEventsQueueOptions } from './log-events-processor'

const isProduction = process.env.NODE_ENV === 'production'
const server = restify.createServer()
const globalThrottleOptions = {
  burst: 3,  // Max 3 concurrent requests
  rate: 1,  // 1 request / second
  ip: true,   // throttle per IP
}

const allowedOrigins = process.env.CORS_ALLOWED_ORIGIN || '*'

const cors = corsMiddleware({
  origins: allowedOrigins.split(','),
})

server.pre(cors.preflight)
server.use(cors.actual)
server.use(restify.plugins.bodyParser())
server.use(restify.plugins.queryParser());
if (isProduction) {
  server.use(restify.plugins.throttle(globalThrottleOptions))
}
const port = process.env.PORT || 8080

const invalidClientTokenErrorBody = {
  error: {
    code: 101,
    message: 'Required parameter "clientToken" is invalid.',
  },
}

const noClientTokenErrorBody = {
  error: {
    code: 100,
    message: 'Required parameter is missing: "clientToken".',
  },
}

const testEndpointRateLimit = restify.plugins.throttle({
  burst: 1,
  rate: 0.015, //  1 request / minute
  ip: true,
})

const testResponse = (req, res, next) => {
  const accessToken = createAccessToken()
  const getRandomInt = max => Math.floor(Math.random() * Math.floor(max))
  const params = {
    accessToken,
    params: {
      MetricData: [
      {
          MetricName: 'HELLO_WORLD',
          Value: getRandomInt(100),
        }
      ],
      Namespace: 'cloudwatch-postman',
    }
  }

  request.post({
    url: `http://localhost:${port}/metric`,
    body: JSON.stringify(params),
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
  }, (error, response, body) => {
    if (error) console.log('>> Error:', error)
    if (response) console.log('>> Status code:', response.statusCode)
    res.json(200, JSON.parse(body))
  })

  return next()
}

if (isProduction) {
  server.get('/test', testEndpointRateLimit, testResponse)
} else {
  server.get('/test', testResponse)
}

server.post('/token', (req, res, next) => {
  if (!req.body || !req.body.clientToken) {
    res.json(403, noClientTokenErrorBody)
    return next()
  }

  const clientToken = req.body.clientToken

  if (!isClientTokenValid(clientToken)) {
    res.json(401, invalidClientTokenErrorBody)
    return next()
  }

  res.json(201, {
    accessToken: createAccessToken(),
  })
  return next()
})

server.post('/metric', (req, res, next) => {
  const { body: { accessToken } } = req

  if (!accessToken) {
    res.json(403, {
      error: {
        code: 110,
        message: 'Required parameter is missing: "accessToken".',
      }
    })
    return next()
  }

  if (!isAccessTokenValid(accessToken)) {
    res.json(401, {
      error: {
        code: 111,
        message: 'Required parameter "accessToken" is invalid.',
      }
    })
    return next()
  }

  const cloudwatch = new AWS.CloudWatch({
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    region: process.env.AWS_REGION,
  })

  cloudwatch.putMetricData(req.body.params, (err, data) => {
    if (err) return console.log(err, err.stack)

    res.json(201, data)
  })

  return next()
})

server.post('/logEvents', (req, res, next) => {
  const { body: { accessToken, logGroupName, logStreamName, logEvents } } = req

  if (!accessToken) {
    res.json(403, {
      error: {
        code: 110,
        message: 'Required parameter is missing: "accessToken".',
      }
    })
    return next()
  }

  if (!isAccessTokenValid(accessToken)) {
    res.json(401, {
      error: {
        code: 111,
        message: 'Required parameter "accessToken" is invalid.',
      }
    })
    return next()
  }

  const logEventsQueue = new Queue('logEvents', logEventsQueueOptions)
  const params = {
    logGroupName,
    logStreamName,
    logEvents,
  }

  logEventsQueue.add(params, {
    attempts: 3,
    removeOnComplete: true,
  })

  res.json(201)

  return next()
})

server.listen(port, () => {
  console.log('%s listening at %s', server.name, server.url)
})

export default server
